types/src/vm/versioning.rs

```diff
@@ -249,6 +251,80 @@ where
     }
 }
 
+/// For values stored in versioned slots, the VM computes size and metadata to charge gas and
+/// process the refund that is derived from the values. This struct implements a data structure
+/// to store such information as metadata / size, with semantics that the data can only be computed
+/// once per version. That is, for some current version V, data can only be computed once. If
+/// fetching derived data for version U > V, it will not be returned as it should be recomputed.
+///
+/// Note: it is the responsibility of the user of [VersionedSlot] and [VersionedSlotDerivedData] to
+/// synchronise access between them. E.g., if derived data is computed from a value, and the value
+/// is further mutated at the same version, the caller needs to decide how to handle such a case.
+pub struct VersionedSlotDerivedData<V> {
+    inner: VersionedSlot<V>,
```

Comment:
why we need an extra struct to track it, can't it be merged with VersionedSlot with an optional field?